listenbrainz/webserver/views/playlist_api.py

```diff
@@ -0,0 +1,516 @@
+    except Exception as e:
+        current_app.logger.error("Error while creating new playlist: {}".format(e))
+        raise APIInternalServerError("Failed to create the playlist. Please try again.")
+
+    return jsonify({'status': 'ok', 'playlist_mbid': playlist.mbid})
+
+
+@playlist_api_bp.route("/<playlist_mbid>", methods=["GET", "OPTIONS"])
+@crossdomain(headers="Authorization, Content-Type")
+@ratelimit()
+def get_playlist(playlist_mbid):
+    """
+    Fetch the given playlist.
+
+    :param playlist_mbid: Optional, The playlist mbid to fetch.
+    :statuscode 200: Yay, you have data!
+    :statuscode 404: Playlist not found
+    :statuscode 401: Invalid authorization. See error message for details.
+    :resheader Content-Type: *application/json*
+    """
+
+    if not is_valid_uuid(playlist_mbid):
+        log_raise_400("Provided playlist ID is invalid.")
+
+    playlist = db_playlist.get_by_mbid(playlist_mbid, True)
+    if playlist is None:
+        raise APINotFound("Cannot find playlist: %s" % playlist_mbid)
+
+    if not playlist.public:
+        user = _validate_auth_header()
+        if playlist.creator_id != user["id"]:
+            raise APINotFound("Cannot find playlist: %s" % playlist_mbid)
+
+    fetch_playlist_recording_metadata(playlist)
+
+    return jsonify(serialize_jspf(playlist))
+
+
+@playlist_api_bp.route("/<playlist_mbid>/item/add/<int:offset>", methods=["POST", "OPTIONS"])
+@playlist_api_bp.route("/<playlist_mbid>/item/add", methods=["POST", "OPTIONS"], defaults={'offset': None})
+@crossdomain(headers="Authorization, Content-Type")
+@ratelimit()
+def add_playlist_item(playlist_mbid, offset):
+    """
+    Append recordings to an existing playlist by posting a playlist with one of more recordings in it.
+    The playlist must be in JSPF format with MusicBrainz extensions, which is defined here:
+    https://musicbrainz.org/doc/jspf .
+
+    If the offset is provided in the URL, then the recordings will be added at that offset,
+    otherwise they will be added at the end of the playlist.
+
+    You may only add :data:`~webserver.views.playlist_api.MAX_RECORDINGS_PER_ADD` recordings in one
+    call to this endpoint.
+
+    :reqheader Authorization: Token <user token>
+    :statuscode 200: playlist accepted.
+    :statuscode 400: invalid JSON sent, see error message for details.
+    :statuscode 401: invalid authorization. See error message for details.
+    :statuscode 403: forbidden. the requesting user was not allowed to carry out this operation.
+    :resheader Content-Type: *application/json*
+    """
+
+    user = _validate_auth_header()
+    if offset is not None and offset < 0:
+        log_raise_400("Offset must be a positive integer.")
+
+    if not is_valid_uuid(playlist_mbid):
+        log_raise_400("Provided playlist ID is invalid.")
+
+    playlist = db_playlist.get_by_mbid(playlist_mbid)
+    if playlist is None or \
+       (playlist.creator_id != user["id"] and not playlist.public):
```

Comment:
new feature: It might be nicer to have a method on playlist that allows us to do this check, e.g. 
`playlist.can_view(user)` and `playlist.can_edit(user)`. It'll help us in the future with collaborators too. We could make this change when we add collab support